src/workerd/api/streams/readable.h

```diff
@@ -342,6 +336,28 @@ class ReadableStream: public jsg::Object {
     // value-oriented streams.
   }
 
+  jsg::Ref<ReadableStream> detach(jsg::Lock& js);
+  // Detaches this ReadableStream from it's underlying controller state, returning a
+  // new ReadableStream instance that takes over the underlying state. This is used to
+  // support the "create a proxy" of a ReadableStream algorithm in the streams spec
+  // (see https://streams.spec.whatwg.org/#readablestream-create-a-proxy). In that
+  // algorithm, it says to create a proxy of a stream by creating a new TransformStream
+  // and piping the original through it. The readable side of the created transform
+  // becomes the proxy. That is quite inefficient so instead, we create a new
+  // ReadableStream that will take over ownership of the internal state of this one,
+  // leaving this ReadableStream locked and disturbed so that it is no longer usable.
+  // The name "detach" here is used in the sense of "detaching the internal state".
+
+  kj::Maybe<uint64_t> tryGetLength(StreamEncoding encoding);
+
+  kj::Promise<DeferredProxy<void>> pumpTo(jsg::Lock& js,
+                                          kj::Own<WritableStreamSink> sink,
+                                          bool end);
+  // A potentially optimized version of pipe that sends this stream's data to the given
+  // sink. The entire stream is consumed. The ReadableStream will be left locked and
+  // disturbed and the DeferredProxy returned will take over ownership of the internal
+  // state of the readable.
```

Comment:
Is the implication of the last sentence that this function is an exception to the KJ convention that callers of promise-returning membrer functions must arrange to prolong the lifetime of the object on which the function is a member?

Reply:
Essentially yes.